File: src/main/java/me/austin/queer/module/gui/clickgui/ClickGuiScreen.java
package me.austin.queer.module.gui.clickgui;

import java.util.ArrayList;
import java.util.List;

import me.austin.queer.module.gui.clickgui.components.CategoryFrame;
import me.austin.queer.module.hacks.Category;
import me.austin.queer.module.hacks.Hacks;
import me.austin.queer.module.hacks.client.ClickGui;
import me.austin.queer.module.setting.Setting;
import me.austin.queer.util.Util;
import me.austin.queer.util.render.ScreenHelper;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.text.LiteralText;

public class ClickGuiScreen extends Screen implements Util {
	public static ClickGuiScreen INSTANCE;

	public final Hacks hacks;
	public List<CategoryFrame> frames;
	public static boolean shouldCloseOnEsc = true;
	
	public ClickGuiScreen(Hacks hacks, Setting<Integer> backgroundRed, Setting<Integer> backgroundGreen, Setting<Integer> backgroundBlue) {
		super(new LiteralText("ClickGuiScreen"));
		this.hacks = hacks;
		int offset = 20 * ScreenHelper.wpixel();

		frames = new ArrayList<>();
		for (Category category : Category.values()) {
			this.frames.add(new CategoryFrame(category, 20 + offset, 20 * ScreenHelper.wpixel(), this));
			offset += 60 * ScreenHelper.wpixel();
		}
		INSTANCE = this;
	}

	@Override
	public void render(MatrixStack matrices, int mousex, int mousey, float delta) {
		for (CategoryFrame frame : frames) {
			frame.render(matrices, mc.textRenderer, frame.x, frame.y);
		}
	}
	
	@Override
	public boolean mouseClicked(double mousex, double mousey, int button) {
		for (CategoryFrame frame : frames) {
			if (ScreenHelper.clickCheck(mousex, mousey, frame.x, frame.y, frame.width, 20)) {
				frame.mouseClicked(mousex, mousey, button);
			}
		}
		mc.openScreen(INSTANCE);
		return super.mouseClicked(mousex, mousey, button);
	}

	@Override 
	public boolean mouseReleased(double mousex, double mousey, int button) {
		for (CategoryFrame frame : frames) {
			if (frame.dragging) {
				frame.dragging = false;
			}
		}
		return super.mouseReleased(mousex, mousey, button);
	} 
	
	@Override
	public boolean isPauseScreen() {
		return false;
	}

	@Override
	public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
		this.frames.forEach(frame -> {
			frame.keyPressed(keyCode, scanCode, modifiers);
		});
		mc.openScreen(INSTANCE);
		return super.keyPressed(keyCode, scanCode, modifiers);
	}

	@Override
	public boolean shouldCloseOnEsc() {
		return shouldCloseOnEsc;
	}

	@Override
	public void renderBackground(MatrixStack matrices, int vOffset) {
		return;
	}

	@Override
	public void onClose() {
		Hacks.getHackByClass(ClickGui.class).disable();
	}
}

File: src/main/kotlin/me/austin/slug/impl/gui/ClickGuiScreen.kt
package me.austin.client.impl.gui

import net.minecraft.client.gui.DrawableHelper
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.text.Text
import me.austin.client.BasicEventManager
import me.austin.client.Slug
import me.austin.client.api.Wrapper
import me.austin.client.api.gui.components.Frame
import me.austin.client.api.gui.components.buttons.Button
import me.austin.client.api.hack.Hack
import me.austin.client.api.hack.HackManager
import me.austin.client.api.Manager
import me.austin.client.events.KeyEvent
import me.austin.rush.listener
import me.austin.client.util.fromJson
import me.austin.client.impl.friend.FriendManager
import me.austin.client.impl.friend.Friend
import java.awt.Color
import java.nio.file.Files
import java.nio.file.Path

object ClickGuiScreen : Screen(Text.of(Slug.NAME)), Wrapper, Manager<Frame<*>, List<Frame<*>>> {
    override val values: List<Frame<*>>

    private val file: Path = Path.of("${Slug.mainDirectory}/clickguiscreen.json")

    private var shouldCloseOnEsc = true

    private var key = 'y'

    private val keyListener = listener<KeyEvent>({
        if (it.key == this.key.code) {
            minecraft.setScreen(ClickGuiScreen)
            it.cancel()
        }
    }, Integer.MAX_VALUE)

    init {
        var offset = 0

        values = listOf(object : Frame<Hack>(20, 20, 60, HackManager.values.size * 20, HackManager.values.map {
            offset += 20

            object : Button<Hack>(20, offset, 60, 20, it) {
                override fun render(stack: MatrixStack) {
                    DrawableHelper.fill(
                        stack,
                        this.xPos,
                        this.yPos,
                        this.width + this.xPos,
                        this.height + this.yPos,
                        Color.LIGHT_GRAY.rgb
                    )
                }
            }
        }) {
            override fun render(stack: MatrixStack) {
                DrawableHelper.fill(
                    stack,
                    this.xPos,
                    this.yPos,
                    this.width + this.xPos,
                    this.height + this.yPos,
                    Color.LIGHT_GRAY.rgb
                )

                for (button in this.children) button.render(stack)
            }
        }, object : Frame<Friend>(100, 20, 60, FriendManager.values.size * 20, FriendManager.values.map {
            offset += 20

            object : Button<Friend>(20, offset, 60, 20, it) {
                override fun render(stack: MatrixStack) {
                    DrawableHelper.fill(
                        stack,
                        this.xPos,
                        this.yPos,
                        this.width + this.xPos,
                        this.height + this.yPos,
                        Color.YELLOW.rgb
                    )
                }
            }
        }) {
            override fun render(stack: MatrixStack) {
                DrawableHelper.fill(
                    stack,
                    this.xPos,
                    this.yPos,
                    this.width + this.xPos,
                    this.height + this.yPos,
                    Color.LIGHT_GRAY.rgb
                )

                for (button in this.children) button.render(stack)
            }
        })
    }

    override fun render(matrices: MatrixStack, mouseX: Int, mouseY: Int, delta: Float) {
        for (frame in values) frame.render(matrices)
    }

    override fun shouldPause() = false

    override fun shouldCloseOnEsc() = shouldCloseOnEsc

    override fun load() {
        if (!Files.exists(file)) Files.createFile(file)

        BasicEventManager.register(keyListener)
    }

    override fun unload() {
        BasicEventManager.unregister(keyListener)
    }
}
